exchanges/kucoin/kucoin_websocket.go

```diff
@@ -0,0 +1,1903 @@
+				Type:     operation,
+				Topic:    fmt.Sprintf(subscriptions[x].Channel, subscriptions[x].Currency.String()),
+				Response: true,
+			}
+			switch subscriptions[x].Channel {
+			case futuresPositionChangeEventChannel,
+				futuresTradeOrdersBySymbolChannel:
+				item.PrivateChannel = true
+			}
+			payloads = append(payloads, item)
+		case marketTickerSnapshotForCurrencyChannel,
+			marginLoanChannel:
+			// 3 means the Currency is used by both switch cases
+			// 2 means the currency is used by channel = marginLoanChannel
+			// 1 if used by marketTickerSnapshotForCurrencyChannel
+			if stat := marketTickerSnapshotForCurrencyChannelCurrencyFilter[subscriptions[x].Currency.Base]; stat == 3 || (stat == 2 && subscriptions[x].Channel == marginLoanChannel) || stat == 1 {
+				continue
+			}
+			input := WsSubscriptionInput{}
+			if subscriptions[x].Channel == marginLoanChannel {
+				input.PrivateChannel = true
+				marketTickerSnapshotForCurrencyChannelCurrencyFilter[subscriptions[x].Currency.Base] += 2
+			} else {
+				marketTickerSnapshotForCurrencyChannelCurrencyFilter[subscriptions[x].Currency.Base]++
+				subscriptions[x].Channel += "%s"
+			}
+			input.ID = strconv.FormatInt(ku.Websocket.Conn.GenerateMessageID(false), 10)
+			input.Type = operation
+			input.Topic = fmt.Sprintf(subscriptions[x].Channel, subscriptions[x].Currency.Base.Upper().String())
+			input.Response = true
+			payloads = append(payloads, input)
+		case marketCandlesChannel:
+			interval, err := ku.intervalToString(subscriptions[x].Params["interval"].(kline.Interval))
+			if err != nil {
+				return nil, err
+			}
+			payloads = append(payloads, WsSubscriptionInput{
+				ID:       strconv.FormatInt(ku.Websocket.Conn.GenerateMessageID(false), 10),
+				Type:     operation,
+				Topic:    fmt.Sprintf(subscriptions[x].Channel, subscriptions[x].Currency.Upper().String(), interval),
+				Response: true,
+			})
+		case marginFundingbookChangeChannel:
+			currencies, okay := subscriptions[x].Params["currencies"].(string)
+			if !okay {
+				return nil, errors.New("currencies not passed")
+			}
+			payloads = append(payloads, WsSubscriptionInput{
+				ID:       strconv.FormatInt(ku.Websocket.Conn.GenerateMessageID(false), 10),
+				Type:     operation,
+				Topic:    fmt.Sprintf(subscriptions[x].Channel, currencies),
+				Response: true,
+			})
+		}
+	}
+	return payloads, nil
+}
+
+// orderbookManager defines a way of managing and maintaining synchronisation
+// across connections and assets.
+type orderbookManager struct {
+	state map[currency.Code]map[currency.Code]map[asset.Item]*update
+	sync.Mutex
+
+	jobs chan job
+}
+
+type update struct {
+	buffer            chan *WsOrderbook
+	fetchingBook      bool
+	initialSync       bool
+	needsFetchingBook bool
+	lastUpdateID      int64
+}
+
+// job defines a synchronisation job that tells a go routine to fetch an
+// orderbook via the REST protocol
+type job struct {
```

Comment:
I'm not sure I have full faith in the orderbook process
Watching the logs I'm getting:
BTC:
```
[INFO]  | SYNC | 21/06/2023 15:59:01 | Kucoin websocket BTC-USDT SPOT ORDERBOOK: Bids len: 106 Amount: 50.722645 BTC. Total value: 1467228.0648536095 Asks len: 102 Amount: 28.046937 BTC. Total value: 812404.581053835

[INFO]  | SYNC | 21/06/2023 15:59:10 | Kucoin websocket BTC-USDT SPOT ORDERBOOK: Bids len: 184 Amount: 71.968982 BTC. Total value: 2082214.2987260348 Asks len: 122 Amount: 36.687630 BTC. Total value: 1062675.2765244981

[INFO]  | SYNC | 21/06/2023 15:59:20 | Kucoin websocket BTC-USDT SPOT ORDERBOOK: Bids len: 208 Amount: 73.469939 BTC. Total value: 2125619.7712647463 Asks len: 135 Amount: 39.329978 BTC. Total value: 1139186.8121617462

[INFO]  | SYNC | 21/06/2023 15:59:30 | Kucoin websocket BTC-USDT SPOT ORDERBOOK: Bids len: 410 Amount: 136.429719 BTC. Total value: 3949437.1551189604 Asks len: 260 Amount: 56.538255 BTC. Total value: 1638200.930026124

[INFO]  | SYNC | 21/06/2023 15:59:40 | Kucoin websocket BTC-USDT SPOT ORDERBOOK: Bids len: 485 Amount: 141.962395 BTC. Total value: 4109549.508040015 Asks len: 315 Amount: 74.513904 BTC. Total value: 2159100.476505304

[INFO]  | SYNC | 21/06/2023 15:59:50 | Kucoin websocket BTC-USDT SPOT ORDERBOOK: Bids len: 531 Amount: 155.443816 BTC. Total value: 4500155.731775174 Asks len: 334 Amount: 85.275431 BTC. Total value: 2471112.601977363

[INFO]  | SYNC | 21/06/2023 16:00:00 | Kucoin websocket BTC-USDT SPOT ORDERBOOK: Bids len: 540 Amount: 147.345144 BTC. Total value: 4265578.295063974 Asks len: 343 Amount: 85.844780 BTC. Total value: 2487628.896811592

[INFO]  | SYNC | 21/06/2023 16:00:10 | Kucoin websocket BTC-USDT SPOT ORDERBOOK: Bids len: 696 Amount: 209.441933 BTC. Total value: 6065910.320754595 Asks len: 502 Amount: 119.312700 BTC. Total value: 3458295.6454338464

[INFO]  | SYNC | 21/06/2023 16:00:20 | Kucoin websocket BTC-USDT SPOT ORDERBOOK: Bids len: 716 Amount: 202.975517 BTC. Total value: 5878547.525321046 Asks len: 529 Amount: 120.072418 BTC. Total value: 3480391.093168332

[INFO]  | SYNC | 21/06/2023 16:00:30 | Kucoin websocket BTC-USDT SPOT ORDERBOOK: Bids len: 722 Amount: 205.816125 BTC. Total value: 5960709.514944135 Asks len: 536 Amount: 121.346354 BTC. Total value: 3517281.1261224616

[INFO]  | SYNC | 21/06/2023 16:00:40 | Kucoin websocket BTC-USDT SPOT ORDERBOOK: Bids len: 740 Amount: 194.746328 BTC. Total value: 5639728.833379179 Asks len: 558 Amount: 127.550591 BTC. Total value: 3697067.6280048746

[INFO]  | SYNC | 21/06/2023 16:00:50 | Kucoin websocket BTC-USDT SPOT ORDERBOOK: Bids len: 768 Amount: 211.584365 BTC. Total value: 6126934.54757863 Asks len: 588 Amount: 143.368654 BTC. Total value: 4155230.575453295
```

ETH

```
[INFO]  | SYNC | 21/06/2023 15:59:01 | Kucoin websocket ETH-USDT MARGIN ORDERBOOK: Bids len: 101 Amount: 503.881250 ETH. Total value: 915821.0345516579 Asks len: 102 Amount: 289.331558 ETH. Total value: 526910.273339161

[INFO]  | SYNC | 21/06/2023 15:59:10 | Kucoin websocket ETH-USDT MARGIN ORDERBOOK: Bids len: 125 Amount: 530.881089 ETH. Total value: 964944.8827720168 Asks len: 116 Amount: 324.415382 ETH. Total value: 590775.3690889

[INFO]  | SYNC | 21/06/2023 15:59:20 | Kucoin websocket ETH-USDT MARGIN ORDERBOOK: Bids len: 135 Amount: 533.752775 ETH. Total value: 970163.6019534689 Asks len: 122 Amount: 332.856913 ETH. Total value: 606140.413989741

[INFO]  | SYNC | 21/06/2023 15:59:31 | Kucoin websocket ETH-USDT MARGIN ORDERBOOK: Bids len: 191 Amount: 797.281741 ETH. Total value: 1449681.7623626576 Asks len: 168 Amount: 319.348745 ETH. Total value: 581547.7596685847

[INFO]  | SYNC | 21/06/2023 15:59:40 | Kucoin websocket ETH-USDT MARGIN ORDERBOOK: Bids len: 210 Amount: 823.438902 ETH. Total value: 1497279.4761558454 Asks len: 178 Amount: 349.951356 ETH. Total value: 637290.1537962258

[INFO]  | SYNC | 21/06/2023 15:59:51 | Kucoin websocket ETH-USDT MARGIN ORDERBOOK: Bids len: 350 Amount: 1426.870434 ETH. Total value: 2596032.565357263 Asks len: 295 Amount: 518.226168 ETH. Total value: 944138.2765444329

[INFO]  | SYNC | 21/06/2023 16:00:01 | Kucoin websocket ETH-USDT MARGIN ORDERBOOK: Bids len: 391 Amount: 1466.318274 ETH. Total value: 2667834.6286246036 Asks len: 321 Amount: 621.926759 ETH. Total value: 1133188.6023573983

[INFO]  | SYNC | 21/06/2023 16:00:11 | Kucoin websocket ETH-USDT MARGIN ORDERBOOK: Bids len: 427 Amount: 1593.419263 ETH. Total value: 2899389.3762539555 Asks len: 352 Amount: 638.342410 ETH. Total value: 1163088.4033297577

[INFO]  | SYNC | 21/06/2023 16:00:20 | Kucoin websocket ETH-USDT MARGIN ORDERBOOK: Bids len: 483 Amount: 1866.451038 ETH. Total value: 3396994.029228442 Asks len: 407 Amount: 606.188172 ETH. Total value: 1104479.0873973588

[INFO]  | SYNC | 21/06/2023 16:00:30 | Kucoin websocket ETH-USDT MARGIN ORDERBOOK: Bids len: 488 Amount: 1887.999107 ETH. Total value: 3436202.593599513 Asks len: 413 Amount: 809.706291 ETH. Total value: 1475613.5799226752

[INFO]  | SYNC | 21/06/2023 16:00:41 | Kucoin websocket ETH-USDT MARGIN ORDERBOOK: Bids len: 506 Amount: 1745.202398 ETH. Total value: 3176185.039772774 Asks len: 424 Amount: 1097.281736 ETH. Total value: 1999906.8314831785

[INFO]  | SYNC | 21/06/2023 16:00:51 | Kucoin websocket ETH-USDT MARGIN ORDERBOOK: Bids len: 526 Amount: 1695.524172 ETH. Total value: 3085657.092908179 Asks len: 442 Amount: 1278.862707 ETH. Total value: 2330725.8070016424
```

I am noting that all values keep accumulating. Are you sure you are [following the process defined here?](https://docs.kucoin.com/#level-2-market-data). I do not see any `== 0` related to orderbooks in this file like it requests

If you are then instead only retrieving [this one instead](https://docs.kucoin.com/#level2-50-best-ask-bid-orders), why are the values beyond 50?

Something is amiss!

Reply:
I tried subscribing with different channels of order book and **_count_** the order book entries. The web-socket connection sends new order book data frequently that's why it is showing a big number.